integrations/snowflake/src/haystack_integrations/components/retrievers/snowflake/snowflake_retriever.py

```diff
@@ -0,0 +1,332 @@
+            database=self.database,
+            db_schema=self.db_schema,
+            warehouse=self.warehouse,
+            login_timeout=self.login_timeout,
+        )
+
+    @classmethod
+    def from_dict(cls, data: Dict[str, Any]) -> "SnowflakeRetriever":
+        """
+        Deserializes the component from a dictionary.
+
+        :param data:
+            Dictionary to deserialize from.
+        :returns:
+           Deserialized component.
+        """
+        init_params = data.get("init_parameters", {})
+        deserialize_secrets_inplace(init_params, ["api_key"])
+        return default_from_dict(cls, data)
+
+    @staticmethod
+    def _snowflake_connector(connect_params: Dict[str, Any]) -> Union[SnowflakeConnection, None]:
+        """
+        Connect to a Snowflake database.
+
+        :param connect_params: Snowflake connection parameters.
+        """
+        try:
+            return snowflake.connector.connect(**connect_params)
+        except DatabaseError as e:
+            logger.error("{error_msg} ", errno=e.errno, error_msg=e.msg)
+            return None
+
+    @staticmethod
+    def _extract_table_names(query: str) -> list:
+        """
+        Extract table names from a SQL query using regex.
+        The extracted table names will be checked for privilege.
+
+        :param query: SQL query to extract table names from.
+        """
+
+        # Regular expressions to match table names in various clauses
+        suffix = "\\s+([a-zA-Z0-9_.]+)"
+
+        patterns = [
+            "MERGE\\s+INTO",
+            "USING",
+            "JOIN",
+            "FROM",
+            "UPDATE",
+            "DROP\\s+TABLE",
+            "TRUNCATE\\s+TABLE",
+            "CREATE\\s+TABLE",
+            "INSERT\\s+INTO",
+            "DELETE\\s+FROM",
+        ]
+
+        # Combine all patterns into a single regex
+        combined_pattern = "|".join([pattern + suffix for pattern in patterns])
+
+        # Find all matches in the query
+        matches = re.findall(pattern=combined_pattern, string=query, flags=re.IGNORECASE)
+
+        # Flatten list of tuples and remove duplication
+        matches = list(set(sum(matches, ())))
+
+        # Clean and return unique table names
+        return [match.strip('`"[]').upper() for match in matches if match]
+
+    @staticmethod
+    def _execute_sql_query(conn: SnowflakeConnection, query: str) -> pd.DataFrame:
+        """
+        Execute a SQL query and fetch the results.
+
+        :param conn: An open connection to Snowflake.
+        :param query: The query to execute.
+        """
+        cur = conn.cursor()
+        try:
+            cur.execute(query)
+            # set a limit to MAX_SYS_ROWS rows to avoid fetching too many rows
+            rows = cur.fetchmany(size=MAX_SYS_ROWS)
+            # Convert data to a dataframe
+            df = pd.DataFrame(rows, columns=[desc.name for desc in cur.description])
+            return df
+        except ProgrammingError as e:
```

Comment:
I'm assuming this is the only way to catch an invalid/erroneous query from the user? If not maybe it's worth to broaden the except so that the user knows that he/she typed an invalid query.